File: 0853-car-fleet/0853-car-fleet.py
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        # approach:
        # 1. get arrival times of each car
        # 2. pair arrival times with start positions
        # 3. sort list of pairs by start position
        # 4. use stack to group car fleets

        length = len(position)
        arrivals = []
        pairs = []
        
        for i in range(length):
            # y = mx + b
            # x = (y - b) / m
            
            y, m, b = target, speed[i], position[i]
            x = (y - b) / m
            
            arrivals.append(x)
            pairs.append((position[i], arrivals[i]))

        pairs.sort(key=lambda x: x[0])

        stack = []
        for i in range(length - 1, -1, -1):
            arrival = pairs[i][1]
            
            if stack and arrival <= stack[-1]:
                continue
            
            stack.append(arrival)
        
        return len(stack)

File: 0853-car-fleet/0853-car-fleet.java
class Solution {
    public int carFleet(int target, int[] position, int[] speed) {
        int n = position.length;
        double[] arrivalTimes = new double[n];
        Stack<Double> stack = new Stack<>();
        
        for (int i = 0; i < n; i++) {
            // y = mx + b
            // x = (y - b) / m
            
            double y = target;
            double m = speed[i];
            double b = position[i];
            
            double x = (y - b) / m;
            
            arrivalTimes[i] = x;
        }
        
        for (double num : arrivalTimes) {
            System.out.println(num);
        }
        
        // array of arrays of size 2 position : arrivalTime
        double[][] cars = new double[n][2];
        
        for (int i = 0; i < n; i++) {
            cars[i][0] = position[i];
            cars[i][1] = arrivalTimes[i];
        }
        
        // sort cars by position
        Arrays.sort(cars, Comparator.comparingDouble(c -> c[0]));
        
        for (double[] car : cars) {
            System.out.println(car[0] + " " + car[1]);
        }
        
        // add arrivalTimes to stack from position closest to target
        for (int i = n - 1; i >= 0; i--) {
            double arrivalTime = cars[i][1];
            
            if (!stack.isEmpty() && arrivalTime <= stack.peek()) {
                continue;
            }
            
            stack.push(arrivalTime);
        }
        
        System.out.println(stack);
                
        return stack.size();
    }
}